File: 06/Help/Beneficiary.java
import java.util.Iterator;
import java.util.ArrayList;

public abstract class Beneficiary implements Iterable {

	private int _receivedHelp = 0;
	private ArrayList<Beneficiary> _parts = new ArrayList<Beneficiary>();

	public Iterator<Beneficiary> iterator() {
		return new Iterator<Beneficiary>() {
			int _index = 0;
			public boolean hasNext() { return _index < size()-1;}
			public Beneficiary next() { return _parts.get(_index++); }
			public void remove() { throw new UnsupportedOperationException(); }
		};
	}

	public void receiveHelp(int help) {
		System.out.println(this + " received " + help + " of help.");
		_receivedHelp += help;
		destributeHelp(help);
	}

	public void destributeHelp(int help) {
		if (size() > 1)
			for (int i = 0; i < size(); ++i) //Beneficiary ben : (Beneficiary)this)
				_parts.get(i).receiveHelp(help/size());
	}

	public int size() { return _parts.size(); }

	public int receivedHelp() { return _receivedHelp; }

	public void addIndividual(Individual ind) {
		_parts.add(ind);
	}

	public void addVillage(Village vil) {
		_parts.add(vil);
	}
}

File: 06_C++/Beneficiary.cpp
// Beneficiary.cpp

#include "Beneficiary.h"
#include <vector>

void Beneficiary::distribute(int help) {
	if (_subdivisions.size() > 0) {
		int help_for_each = help / _subdivisions.size();

		if (help_for_each > 0) {
			std::vector<Beneficiary*>::iterator it;
			for(it = _subdivisions.begin() ; it < _subdivisions.end(); it++)
				(*it)->receive_help(help_for_each);
		}
	}
}

int Beneficiary::total_subdivisions_received_help() const {
	if (_subdivisions.size() == 0)
		return _received_help;

	int total = 0;

	std::vector<Beneficiary*>::const_iterator it;
	for(it = _subdivisions.begin() ; it < _subdivisions.end(); it++)
		total += (*it)->total_subdivisions_received_help();
	
	return total;
}	